### PR TITLE
[NOT FOR MERGE] Exploratory work on AOTInductor training

### PR DESCRIPTION
**Pull request by benjaminglass1**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #155877
* #158074

TODO for MVP:

- [x] Select between training and eval mode within the export model loaders.
- [ ] Improve UI around parameter updates and gradients (i.e. users should not have to manually apply gradients and maintain a separate copy of the model to apply them to for this to work).
- [x] Determine why some tensor constants get dropped when doing AOTInductor training.
- [x] Ensure optimizer runtime is integrated into export performance stats. :( Maybe we can export the full `forward_and_backward_pass` function?
- [x] Break out unrelated code improvements into a separate PR.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

File: torch/_inductor/__init__.py [hunk 1]
```diff
@@ -6,13 +6,13 @@
...
 from typing import Any, IO, Literal, Optional, TYPE_CHECKING, Union
 
-import torch._inductor.config
 import torch.fx
 
 from .standalone_compile import CompiledArtifact  # noqa: TC001
 
 
 if TYPE_CHECKING:
+    from torch._inductor.package import AOTICompiledModel
```
Comment (benjaminglass1) [inline torch/_inductor/__init__.py]: Import this from export instead.

File: benchmarks/dynamo/torchbench.py [hunk 2]
```diff
@@ -464,14 +463,13 @@ def forward_and_backward_pass(self, mod, inputs, collect_outputs=True):
...
         with self.autocast(**self.autocast_arg):
             if isinstance(cloned_inputs, dict):
-                pred = mod(**cloned_inputs)
+                loss, *pred = mod(**cloned_inputs)
             else:
-                pred = mod(*cloned_inputs)
-            loss = self.compute_loss(pred)
-        self.grad_scaler.scale(loss).backward()
+                loss, *pred = mod(*cloned_inputs)
+            self.grad_scaler.scale(loss).backward()
```
Comment (benjaminglass1) [inline benchmarks/dynamo/torchbench.py]: Remove this from autocast again.

File: benchmarks/dynamo/torchbench.py [hunk 3]
```diff
@@ -464,14 +463,13 @@ def forward_and_backward_pass(self, mod, inputs, collect_outputs=True):
...
         self.optimizer_step()
         if collect_outputs:
-            return collect_results(mod, None, loss, cloned_inputs)
+            return collect_results(mod, pred, loss, cloned_inputs)
```
Comment (benjaminglass1) [inline benchmarks/dynamo/torchbench.py]: Revert this?
